File: src/main/java/net/ehicks/bts/beans/Issue.java
//package net.ehicks.bts.beans;
//
//import net.ehicks.eoi.EOI;
//import net.ehicks.eoi.Index;
//import net.ehicks.eoi.Indexes;
//
//import javax.persistence.*;
//import java.io.Serializable;
//import java.time.LocalDateTime;
//import java.time.temporal.ChronoUnit;
//import java.util.Arrays;
//import java.util.List;
//
//@Entity
//@Table(name = "issues")
//@Indexes({
//        @Index(sql = "CREATE INDEX  ON ISSUES USING gin (title gin_trgm_ops);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES USING gin (description gin_trgm_ops);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(title);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(description);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(group_id);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(issue_type_id);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(project_id);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(assignee_user_id);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(reporter_user_id);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(severity_id);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(created_on);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(last_updated_on);"),
//        @Index(sql = "CREATE INDEX  ON ISSUES(status_id);")
//})
//public class Issue implements Serializable
//{
//    @Id
//    @GeneratedValue(strategy=GenerationType.AUTO)
//    private Long id;
//
//    @Column(length = 2000)
//    private String title = "";
//    @Column(columnDefinition = "varchar2(32000 CHAR)")
//    private String description = "";
//
//    private Long groupId;
//    private Long issueTypeId;
//    private Long projectId;
//    private Long assigneeUserId;
//    private Long reporterUserId;
//    private Long severityId;
//    private Long statusId;
//
//    @Override
//    public boolean equals(Object obj)
//    {
//        if (!(obj instanceof Issue)) return false;
//        Issue that = (Issue) obj;
//        return this.id.equals(that.getId());
//    }
//
//    @Override
//    public int hashCode()
//    {
//        return 17 * 37 * id.intValue();
//    }
//
//    public String toString()
//    {
//        return this.getClass().getSimpleName() + ":" + id.toString();
//    }
//
//    // --------
//
//    public static List<Issue> getAll()
//    {
//        return EOI.executeQuery("select * from issues");
//    }
//
//    public static Issue getById(Long id)
//    {
//        return EOI.executeQueryOneResult("select * from issues where id=?", Arrays.asList(id));
//    }
//
//    public Project getProject()
//    {
//        return Project.getById(projectId);
//    }
//
//    public Group getGroup()
//    {
//        return Group.getById(groupId);
//    }
//
//    public IssueType getIssueType()
//    {
//        return IssueType.getById(issueTypeId);
//    }
//
//    public Severity getSeverity()
//    {
//        return Severity.getById(severityId);
//    }
//
//    public Status getStatus()
//    {
//        return Status.getById(statusId);
//    }
//
//    public User getAssignee()
//    {
//        return User.getByUserId(assigneeUserId);
//    }
//
//    public User getReporter()
//    {
//        return User.getByUserId(reporterUserId);
//    }
//
//    public List<Attachment> getAttachments()
//    {
//        return Attachment.getByIssueId(id);
//    }
//
//    public String getTimeSinceCreation()
//    {
////        LocalDateTime start = LocalDateTime.ofInstant(Instant.ofEpochMilli(this.getCreatedOn().getTime()), ZoneId.systemDefault());
//        LocalDateTime start = LocalDateTime.now();
//        LocalDateTime until = LocalDateTime.now();
//
//        long days = ChronoUnit.DAYS.between(start, until);
//        long hours = ChronoUnit.HOURS.between(start, until);
//        long minutes = ChronoUnit.MINUTES.between(start, until);
//
//        if (days > 0)
//            return days + " days";
//        if (hours > 0)
//            return hours + " hours";
//        if (minutes > 0)
//            return minutes + " minutes";
//
//        return "<1 minute";
//    }
//    // -------- Getters / Setters ----------
//
//
//    public Long getId()
//    {
//        return id;
//    }
//
//    public void setId(Long id)
//    {
//        this.id = id;
//    }
//
//    public String getTitle()
//    {
//        return title;
//    }
//
//    public void setTitle(String title)
//    {
//        this.title = title;
//    }
//
//    public String getDescription()
//    {
//        return description;
//    }
//
//    public void setDescription(String description)
//    {
//        this.description = description;
//    }
//
//    public Long getGroupId()
//    {
//        return groupId;
//    }
//
//    public void setGroupId(Long groupId)
//    {
//        this.groupId = groupId;
//    }
//
//    public Long getIssueTypeId()
//    {
//        return issueTypeId;
//    }
//
//    public void setIssueTypeId(Long issueTypeId)
//    {
//        this.issueTypeId = issueTypeId;
//    }
//
//    public Long getProjectId()
//    {
//        return projectId;
//    }
//
//    public void setProjectId(Long projectId)
//    {
//        this.projectId = projectId;
//    }
//
//    public Long getAssigneeUserId()
//    {
//        return assigneeUserId;
//    }
//
//    public void setAssigneeUserId(Long assigneeUserId)
//    {
//        this.assigneeUserId = assigneeUserId;
//    }
//
//    public Long getReporterUserId()
//    {
//        return reporterUserId;
//    }
//
//    public void setReporterUserId(Long reporterUserId)
//    {
//        this.reporterUserId = reporterUserId;
//    }
//
//    public Long getSeverityId()
//    {
//        return severityId;
//    }
//
//    public void setSeverityId(Long severityId)
//    {
//        this.severityId = severityId;
//    }
//
//    public Long getStatusId()
//    {
//        return statusId;
//    }
//
//    public void setStatusId(Long statusId)
//    {
//        this.statusId = statusId;
//    }
//}

File: src/main/java/net/ehicks/bts/beans/Issue.kt
package net.ehicks.bts.beans

import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import java.io.Serializable
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import javax.persistence.*

@Entity
@Table(indexes = [
    Index(name = "IDX_ISSUE_TITLE", columnList = "title"),
    Index(name = "IDX_ISSUE_DESCRIPTION", columnList = "description"),
    Index(name = "IDX_ISSUE_GROUP_ID", columnList = "group_id"),
    Index(name = "IDX_ISSUE_ISSUE_TYPE_ID", columnList = "issue_type_id"),
    Index(name = "IDX_ISSUE_PROJECT_ID", columnList = "project_id"),
    Index(name = "IDX_ISSUE_ASSIGNEE_USER_ID", columnList = "assignee_id"),
    Index(name = "IDX_ISSUE_REPORTER_USER_ID", columnList = "reporter_id"),
    Index(name = "IDX_ISSUE_SEVERITY_ID", columnList = "severity_id"),
    Index(name = "IDX_ISSUE_STATUS_ID", columnList = "status_id")
])
data class Issue @JvmOverloads constructor(
        @Id @GeneratedValue(strategy = GenerationType.AUTO)
        var id: Long = 0,
        @Column(length = 2000)
        var title: String = "",
        @Column(length=10485760)
        var description: String = "",
        var createdOn: LocalDateTime = LocalDateTime.now(),
        @LastModifiedDate
        var lastUpdatedOn: LocalDateTime = LocalDateTime.now(),

        @ManyToOne(fetch = FetchType.LAZY) var group: Group,
        @ManyToOne(fetch = FetchType.LAZY) var issueType: IssueType,
        @ManyToOne(fetch = FetchType.LAZY) var project: Project,
        @ManyToOne(fetch = FetchType.LAZY) var assignee: User,
        @ManyToOne(fetch = FetchType.LAZY) var reporter: User,
        @ManyToOne(fetch = FetchType.LAZY) var severity: Severity,
        @ManyToOne(fetch = FetchType.LAZY) var status: Status,

        @ManyToMany
        @JoinTable(name = "issue_watcher",
                joinColumns = [JoinColumn(name = "issue_id")],
                inverseJoinColumns = [JoinColumn(name = "watcher_id")])
        var watchers: Set<User> = HashSet()
) : Serializable {
    @OneToMany(mappedBy = "issue")
    @OrderBy("createdOn, id asc")
    var attachments: Set<Attachment> = HashSet()
    @OneToMany(mappedBy = "issue")
    @OrderBy("createdOn, id asc")
    var comments: Set<Comment> = HashSet()

    fun getTimeSinceCreation(): String {
        return getTimeSince(createdOn)
    }

    fun getTimeSinceUpdate(): String {
        return getTimeSince(lastUpdatedOn)
    }

    val key: String
        get() = project.prefix + "-" + id

    private fun getTimeSince(since: LocalDateTime): String {
        val now: LocalDateTime = LocalDateTime.now()
        val chronos = ChronoUnit.values()
                .slice(ChronoUnit.MILLIS.ordinal..ChronoUnit.YEARS.ordinal)
                .filterNot { it == ChronoUnit.HALF_DAYS }
                .reversed()
        for (chrono in chronos)
        {
            val amount = chrono.between(since, now)
            if (amount > 0)
            {
                var unit = chrono.toString()
                if (amount == 1L) unit = unit.dropLast(1)
                return "$amount $unit"
            }
        }
        return "undefined"
    }
}

@Repository
interface IssueRepository : JpaRepository<Issue, Long> {
    fun findFirstByOrderById(): Issue
}